### ISSUE TITLE
Adding pipeline with scope single and input_output is a dict crashes metadata_lib

### ISSUE THREAD

**Issue by jeroenkroesen** [jeroenkroesen/metadata_lib#2]
This issue was reported by the data engineering team at Dedimo.

Use metadata_manager to add this new pipeline:
```python
new_pipeline = Pipeline(
    id=get_next_free_id(mm.workspace.by_id, 'pipelines'),
    unid=None,
    namespace='dedimo',
    name='testest',
    description='Ingest pipeline for freshservice',
    enabled=True,
    version=1,
    scope='compound',
    type='ingest',
    velocity='batch',
    input_output={'input':'adhdcentraal.adhdcentraal.medicore.risks_conditions.datasource', 'output':'adhdcentraal.dataplatform.google_bigquery.risks_conditions.dataset'},
    config={'apiversion': 'beta'},
    created=pendulum.now(tz='Europe/Amsterdam'),
    modified=pendulum.now(tz='Europe/Amsterdam')
)
print(new_pipeline)
```

Causes this error:
```traceback
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 mm.add_new_entity(new_pipeline)

File [~/02-local/dev/Dedimo/dataplatform_repos/metadata/metadata_manager/env/lib/python3.10/site-packages/metadata_lib/manage.py:498](http://localhost:8888/env/lib/python3.10/site-packages/metadata_lib/manage.py#line=497), in MetadataManager.add_new_entity(self, entity)
    491     result = validate_new_data_entity(
    492         data_entity=entity,
    493         metadata_obj=self.workspace.md_objects,
    494         metadata_id_idx=self.workspace.by_id,
    495         metadata_unid_idx=self.workspace.by_unid
    496     )
    497 elif isinstance(entity, Pipeline):
--> 498     result = validate_new_pipeline(
    499         pipeline=entity,
    500         metadata_obj=self.workspace.md_objects,
    501         metadata_id_idx=self.workspace.by_id,
    502         metadata_unid_idx=self.workspace.by_unid
    503     )
    504 else:
    505     ent_type = type(entity).__name__

File [~/02-local/dev/Dedimo/dataplatform_repos/metadata/metadata_manager/env/lib/python3.10/site-packages/metadata_lib/validate.py:869](http://localhost:8888/env/lib/python3.10/site-packages/metadata_lib/validate.py#line=868), in validate_new_pipeline(pipeline, metadata_obj, metadata_id_idx, metadata_unid_idx)
    833 """Run all checks against a given pipeline
    834 
    835 Perform the following validations against the given pipeline:
   (...)
    866     If references to objects are not ids or unids (int or str)
    867 """
    868 # Run standard validations
--> 869 result = validate_pipeline(
    870     pipeline,
    871     metadata_obj,
    872     metadata_id_idx,
    873     metadata_unid_idx
    874 )
    875 if not result['valid']:
    876     # Errors already. Return without further tests
    877     return result

File [~/02-local/dev/Dedimo/dataplatform_repos/metadata/metadata_manager/env/lib/python3.10/site-packages/metadata_lib/validate.py:449](http://localhost:8888/env/lib/python3.10/site-packages/metadata_lib/validate.py#line=448), in validate_pipeline(pipeline, metadata_obj, metadata_id_idx, metadata_unid_idx)
    445 # Entities referenced in input_output exist
    446 # Iterate instances
    447 for instance in pipeline.input_output:
    448     # Proces input
--> 449     if isinstance(instance['input'], list):
    450         # List style: iterate multiple entities
    451         for inent in instance['input']:
    452             if isinstance(inent, int):

TypeError: string indices must be integers
```

**Comment by jeroenkroesen** [comment]
Fixed above exception in [5da96b959f45060c8e9e321f5a677634c990398a](https://github.com/jeroenkroesen/metadata_lib/commit/5da96b959f45060c8e9e321f5a677634c990398a.

New exception occurs:
```traceback
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 mm.add_new_entity(new_pipeline)

File [~/02-local/dev/Dedimo/dataplatform_repos/metadata/metadata_manager/env/lib/python3.10/site-packages/metadata_lib/manage.py:498](http://localhost:8888/env/lib/python3.10/site-packages/metadata_lib/manage.py#line=497), in MetadataManager.add_new_entity(self, entity)
    491     result = validate_new_data_entity(
    492         data_entity=entity,
    493         metadata_obj=self.workspace.md_objects,
    494         metadata_id_idx=self.workspace.by_id,
    495         metadata_unid_idx=self.workspace.by_unid
    496     )
    497 elif isinstance(entity, Pipeline):
--> 498     result = validate_new_pipeline(
    499         pipeline=entity,
    500         metadata_obj=self.workspace.md_objects,
    501         metadata_id_idx=self.workspace.by_id,
    502         metadata_unid_idx=self.workspace.by_unid
    503     )
    504 else:
    505     ent_type = type(entity).__name__

File [~/02-local/dev/Dedimo/dataplatform_repos/metadata/metadata_manager/env/lib/python3.10/site-packages/metadata_lib/validate.py:964](http://localhost:8888/env/lib/python3.10/site-packages/metadata_lib/validate.py#line=963), in validate_new_pipeline(pipeline, metadata_obj, metadata_id_idx, metadata_unid_idx)
    958     result['errors'].append(
    959         f'A pipeline with id {pipeline.id} exists already'
    960     )
    961 # Unique UNID:
    962 ## Rebuild id_idx with new data_entity included to be able to use  
    963 ## unid generation functions.
--> 964 tmp_pl = unid_refs_to_ids(
    965     metadata_unid_idx=metadata_unid_idx,
    966     obj=deepcopy(pipeline)
    967 )
    968 tmp_obj = deepcopy(metadata_obj)
    969 tmp_obj['pipelines'].append(tmp_pl)

File [~/02-local/dev/Dedimo/dataplatform_repos/metadata/metadata_manager/env/lib/python3.10/site-packages/metadata_lib/build.py:809](http://localhost:8888/env/lib/python3.10/site-packages/metadata_lib/build.py#line=808), in unid_refs_to_ids(metadata_unid_idx, obj)
    807     obj = data_entity_unid_refs_to_ids(metadata_unid_idx, obj)
    808 elif isinstance(obj, Pipeline):
--> 809     obj = pipeline_unid_refs_to_ids(metadata_unid_idx, obj)
    810 return obj

File [~/02-local/dev/Dedimo/dataplatform_repos/metadata/metadata_manager/env/lib/python3.10/site-packages/metadata_lib/build.py:761](http://localhost:8888/env/lib/python3.10/site-packages/metadata_lib/build.py#line=760), in pipeline_unid_refs_to_ids(metadata_unid_idx, pipeline)
    759 # input_output
    760 for i, instance in enumerate(pipeline.input_output):
--> 761     if isinstance(instance['input'], list):
    762         # Multiple input entities. Iterate
    763         for j, inent in enumerate(instance['input']):
    764             if isinstance(inent, str):

TypeError: string indices must be integers
```

**Comment by jeroenkroesen** [comment]
Fixed above exception. Single scope pipelines with dict input_output now validate. 

New exception when choosing yes to add it to workspace:
```traceback
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 mm.add_new_entity(new_pipeline)

File [~/02-local/dev/Dedimo/dataplatform_repos/metadata/metadata_manager/env/lib/python3.10/site-packages/metadata_lib/manage.py:555](http://localhost:8888/lab/tree/env/lib/python3.10/site-packages/metadata_lib/manage.py#line=554), in MetadataManager.add_new_entity(self, entity)
    553     self.cache.md_objects['pipelines'].sort(key=lambda x: x.id)
    554 # Rebuild cache from its object structure, which also validates it.
--> 555 self.cache.build_from_objects()
    556 # Check if cache is still a valid structure
    557 if self.cache.valid:

File [~/02-local/dev/Dedimo/dataplatform_repos/metadata/metadata_manager/env/lib/python3.10/site-packages/metadata_lib/manage.py:179](http://localhost:8888/lab/tree/env/lib/python3.10/site-packages/metadata_lib/manage.py#line=178), in MetadataStructure.build_from_objects(self)
    177 self.md_obj_with_unids = generate_unids(self.by_id)
    178 self.by_unid = create_unid_indexes(self.md_obj_with_unids)
--> 179 self.integrated = integrate_pipelines(deepcopy(self.by_id))
    180 self.dag_config = integrated_to_dag_config(deepcopy(self.integrated))
    181 unidseverywhere = all_ids_to_unids(deepcopy(self.by_id))

File [~/02-local/dev/Dedimo/dataplatform_repos/metadata/metadata_manager/env/lib/python3.10/site-packages/metadata_lib/build.py:416](http://localhost:8888/lab/tree/env/lib/python3.10/site-packages/metadata_lib/build.py#line=415), in integrate_pipelines(metadata_id_idx, flatten_namespaces, flatten_schemas)
    413 # Iterate over this pipelines' instances
    414 for i, instance in enumerate(pl.input_output):
    415     ## Connect entity objects to their integer id in inputs
--> 416     if isinstance(instance['input'], list):
    417         # List style input
    418         for j, inent in enumerate(instance['input']):
    419             pl.input_output[i]['input'][j] = ent_byid[inent]

TypeError: string indices must be integers
```
